[AutoDiff] Relax `@differentiable` requirement for protocol witnesses.

dan-zheng opened the pull request:
Previously, all witnesses of a `@differentiable` protocol requirement were
required to have the same attribute (or one with superset parameter indices).

However, this leads to many annotations on witnesses and is not ideal for
usability. `@differentiable` attributes are really only significant on
public witnesses, so that they are clearly `@differentiable` at a glance (in
source code, interface files, and API documentation), without looking through
protocol conformance hierarchies.

Now, only *public* witnesses of `@differentiable` protocol requirements are
required to have the same attribute (or one with superset parameter indices).
For less-visible witnesses, an implicit `@differentiable` attribute is created
with the same configuration as the requirement's.

Resolves TF-1117.

---

This usability improvement was discussed during the
[2020-01-17 Swift for TensorFlow open design review](https://docs.google.com/document/d/1Fm56p5rV1t2Euh6WLtBFKGqI43ozC3EIjReyLk-LCLU/edit#bookmark=id.rrfb0uhvuzry).
We agreed that it's a good idea!

Todo: upstream changes to `master`.

---

Example:
```swift
public protocol Layer: Differentiable {
  associatedtype Input: Differentiable
  associatedtype Output: Differentiable
  @differentiable(wrt: (self, input))
  func callAsFunction(_ input: Input) -> Output
}

// Internal conforming type.
struct DummyInternalLayer: Layer {
  func callAsFunction(_ input: Float) -> Float {
    return input
  }
}
```

Before (misleading diagnostic due to TF-1014):
```console
layer.swift:9:8: error: type 'DummyInternalLayer' does not conform to protocol 'Layer'
struct DummyInternalLayer: Layer {
       ^
layer.swift:2:18: note: protocol requires nested type 'Input'; do you want to add it?
  associatedtype Input: Differentiable
                 ^
layer.swift:3:18: note: protocol requires nested type 'Output'; do you want to add it?
  associatedtype Output: Differentiable
                 ^
```

After: no error.

An implicit `@differentiable` attribute is created for `DummyInternalLayer.callAsFunction`:
```console
$ swiftc -print-ast layer.swift
...

internal struct DummyInternalLayer : Layer {
  @differentiable(wrt: (self, input))
  internal func callAsFunction(_ input: Float) -> Float
  ...
}
```

dan-zheng committed:
[AutoDiff] Relax `@differentiable` requirement for protocol witnesses.

Previously, all witnesses of a `@differentiable` protocol requirement were
required to have the same attribute (or one with superset parameter indices).

However, this leads to many annotations on witnesses and is not ideal for
usability. `@differentiable` attributes are really only significant on
public witnesses, so that they are clearly `@differentiable` at a glance (in
source code, interface files, and API documentation), without looking through
protocol conformance hierarchies.

Now, only *public* witnesses of `@differentiable` protocol requirements are
required to have the same attribute (or one with superset parameter indices).
For less-visible witnesses, an implicit `@differentiable` attribute is created
with the same configuration as the requirement's.

Resolves TF-1117.

rxwei reviewed (approved):
This makes the implementation consistent with #29307. LGTM.

dan-zheng commented:
@swift-ci Please test tensorflow

rxwei reviewed (commented):
One thing: Creating implicit attributes leads to invalid source locations.  Have you handled such `@differentiable` attributes differentiation diagnostics?  In the transform, the differentiation invoker will be `DifferentiableAttribute`, and any diagnostics will try to get the source location of the implicit attribute.

marcrasi commented:
I thought of a possible weird edge case. What if the conformance is declared in a separate file from the file where the function satisfying the requirement is declared? Then the differentiation pass on the file with the function never sees the implicit differentiable attribute because it only exists in the separate compilation where the conformance is declared.

This edge case probably also applies to the existing subset logic, so it's probably an existing problem that can be dealt with later.

rxwei commented:
@marcrasi Interesting case. One possible solution is to disallow any implicit attribute inheritance when the witness is in a separate translation unit. We’ll just emit an error in protocol witness matching, stating which attribute the potential witness is missing.

dan-zheng committed:
Address review feedback.

Do not generate implicit `@differentiable` attribute on protocol witness when
conformance is in a different file.

Gardening. Add tests.

dan-zheng commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -384,15 +384,25 @@ matchWitnessDifferentiableAttr(DeclContext *dc, ValueDecl *req,
+      //     `@differentiable` attributes for less-visible declarations.
+      bool createImplicitWitnessAttribute =
+          supersetConfig || witness->getFormalAccess() < AccessLevel::Public;
+      if (supersetConfig || witness->getFormalAccess() < AccessLevel::Public) {
```
Done.

dan-zheng commented:
> I thought of a possible weird edge case. What if the conformance is declared in a separate file from the file where the function satisfying the requirement is declared? Then the differentiation pass on the file with the function never sees the implicit differentiable attribute because it only exists in the separate compilation where the conformance is declared.

I [added cross-file tests](https://github.com/apple/swift/blob/31797e0212a1ceb1921064a11462bede7ac448ce/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/main.swift) for this edge case:
- [`main.swift`](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/main.swift)
- [`other_file.swift`](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/Inputs/other_file.swift)

Indeed, implicit `@differentiable` attributes cannot be generated for protocol witnesses when the conformance is declared from a separate file from the witnesses. Otherwise, compilation of the file containing the conformance creates external references to symbols for implicit `@differentiable` attributes, even though no such symbols exist.

I implemented [@rxwei's solution](https://github.com/apple/swift/pull/29771#issuecomment-585217384): a diagnostic is emitted for these cases.

---

> One thing: Creating implicit attributes leads to invalid source locations. Have you handled such `@differentiable` attributes differentiation diagnostics? In the transform, the differentiation invoker will be `DifferentiableAttribute`, and any diagnostics will try to get the source location of the implicit attribute.

Implicit protocol witness `@differentiable` attributes are actually created with the source location from the protocol requirement's `@differentiable` attribute.

I added [a test](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_sil.swift) to verify this:
```swift
protocol Protocol: Differentiable {
  @differentiable(wrt: (self, x))
  func internalMethod(_ x: Float) -> Float
}

struct ConformingStruct: Protocol {
  // Non-differentiable body!
  func internalMethod(_ x: Float) -> Float {
    return Float(Int(x))
  }
}
```
```
test.swift:2:4: error: function is not differentiable
  @differentiable(wrt: (self, x))
  ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.swift:8:8: note: when differentiating this function definition
  func internalMethod(_ x: Float) -> Float {
       ^
test.swift:9:18: note: cannot differentiate through a non-differentiable result; do you want to use 'withoutDerivative(at:)'?
    return Float(Int(x))
                 ^
```

I think these source locations are reasonable. Let me know if you have better suggestions.

---

Ready for re-review! Changes will be upstreamed to `master`.

dan-zheng commented on lib/Sema/TypeCheckProtocol.cpp:
```diff
@@ -384,15 +385,29 @@ matchWitnessDifferentiableAttr(DeclContext *dc, ValueDecl *req,
+      bool createImplicitWitnessAttribute =
+          supersetConfig ||
+          (witness->getFormalAccess() < AccessLevel::Public &&
+           dc->getModuleScopeContext() ==
```
Note: we could add a more specific diagnostic for "witness has less-than-public visibility but is declared in a different file than the conformance".

We need a good diagnostic message though. Suggestions welcome:
> note: candidate is missing attribute '@differentiable'; explicit '@differentiable' attribute is required for non-public declaration 'foo' because it is in a different file than the 'ConformingType: Protocol' conformance

rxwei replied:
Feels like you've squashed two different errors into one. Instead, could you use one error for _non-public declarations in other files_, and a different error for _public declarations_?

dan-zheng replied:
Done in https://github.com/apple/swift/pull/29771/commits/e22a4e8914c233a3b83e718bd093a26a7d4dbd74.

Here's the dedicated diagnostic for "non-public declarations in other files":
```console
main.swift:28:1: error: type 'ConformingStruct' does not conform to protocol 'Protocol'
extension ConformingStruct: Protocol {}
^
Inputs/other_file.swift:33:8: note: non-public instance method 'internalMethod1' must have explicit '@differentiable' attribute to satisfy requirement instance method 'internalMethod1' (in protocol 'Protocol') because it is declared in a different file than the conformance of 'ConformingStruct' to 'Protocol'
  func internalMethod1(_ x: Float) -> Float {
       ^
  @differentiable
```
> non-public instance method 'internalMethod1' must have explicit '@differentiable' attribute to satisfy requirement instance method 'internalMethod1' (in protocol 'Protocol') because it is declared in a different file than the conformance of 'ConformingStruct' to 'Protocol'

It's pretty long but accurate. I copied the template from other witness/requirement mismatch diagnostics.

rxwei commented:
> > I thought of a possible weird edge case. What if the conformance is declared in a separate file from the file where the function satisfying the requirement is declared? Then the differentiation pass on the file with the function never sees the implicit differentiable attribute because it only exists in the separate compilation where the conformance is declared.
> 
> I [added cross-file tests](https://github.com/apple/swift/blob/31797e0212a1ceb1921064a11462bede7ac448ce/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/main.swift) for this edge case:
> 
> * [`main.swift`](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/main.swift)
> * [`other_file.swift`](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_cross_file/Inputs/other_file.swift)
> 
> Indeed, implicit `@differentiable` attributes cannot be generated for protocol witnesses when the conformance is declared from a separate file from the witnesses. Otherwise, compilation of the file containing the conformance creates external references to symbols for implicit `@differentiable` attributes, even though no such symbols exist.
> 
> I implemented [@rxwei's solution](https://github.com/apple/swift/pull/29771#issuecomment-585217384): a diagnostic is emitted for these cases.
> 
> > One thing: Creating implicit attributes leads to invalid source locations. Have you handled such `@differentiable` attributes differentiation diagnostics? In the transform, the differentiation invoker will be `DifferentiableAttribute`, and any diagnostics will try to get the source location of the implicit attribute.
> 
> Implicit protocol witness `@differentiable` attributes are actually created with the source location from the protocol requirement's `@differentiable` attribute.
> 
> I added [a test](https://github.com/apple/swift/blob/4415d5839bc07591b0eeca91bef41a5cb65fa819/test/AutoDiff/downstream/implicit_nonpublic_differentiable_attr_sil.swift) to verify this:
> 
> ```swift
> protocol Protocol: Differentiable {
>   @differentiable(wrt: (self, x))
>   func internalMethod(_ x: Float) -> Float
> }
> 
> struct ConformingStruct: Protocol {
>   // Non-differentiable body!
>   func internalMethod(_ x: Float) -> Float {
>     return Float(Int(x))
>   }
> }
> ```
> 
> ```
> test.swift:2:4: error: function is not differentiable
>   @differentiable(wrt: (self, x))
>   ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> test.swift:8:8: note: when differentiating this function definition
>   func internalMethod(_ x: Float) -> Float {
>        ^
> test.swift:9:18: note: cannot differentiate through a non-differentiable result; do you want to use 'withoutDerivative(at:)'?
>     return Float(Int(x))
>                  ^
> ```
> 
> I think these source locations are reasonable. Let me know if you have better suggestions.
> 
> Ready for re-review! Changes will be upstreamed to `master`.

It is great that we don't have invalid source locations, but this error message could be improved -- it's currently showing an error **on the protocol definition** as if the protocol was defined wrong, which is misleading.  Instead, we should show an error on the declaration that actually triggers the error, along with a note that points us to the protocol requirement that had the `@differentiable` attribute.

dan-zheng committed:
Add dedicated diagnostic: "missing non-public `@differentiable` attribute".

Add dedicated diagnostic for missing `@differentiable` attribute for non-public
protocol witness declared in a different file than the conformance.

dan-zheng committed:
Improve implicit `@differentiable` attribute source location.

Use witness declaration's source location instead of protocol requirement
attribute's source location.

dan-zheng commented:
> It is great that we don't have invalid source locations, but this error message could be improved -- it's currently showing an error **on the protocol definition** as if the protocol was defined wrong, which is misleading. Instead, we should show an error on the declaration that actually triggers the error, along with a note that points us to the protocol requirement that had the `@differentiable` attribute.

Done in https://github.com/apple/swift/pull/29771/commits/03a22e5f1d1a86e780448cfa591c1458d291be1d:
```console
test.swift:20:8: error: function is not differentiable
  func internalMethod(_ x: Float) -> Float {
       ^~~~~~~~~~~~~~
test.swift:20:8: note: when differentiating this function definition
  func internalMethod(_ x: Float) -> Float {
       ^
test.swift:22:18: note: cannot differentiate through a non-differentiable result; do you want to use 'withoutDerivative(at:)'?
    return Float(Int(x))
                 ^
```
Ready for re-review.

rxwei commented:
Hmm the updated implementation is not showing a note on the protocol requirement that had the attribute.

dan-zheng commented:
> Hmm the updated implementation is not showing a note on the protocol requirement that had the attribute.

Done:
```console
test.swift:17:15: error: type 'PublicConformingStruct' does not conform to protocol 'InternalProtocol'
public struct PublicConformingStruct: InternalProtocol {
              ^
test.swift:20:15: note: candidate is missing attribute '@differentiable'
  public func publicMethod(_ x: Float) -> Float {
              ^
         @differentiable
test.swift:8:4: note: required '@differentiable' attribute declared here
  @differentiable(wrt: (self, x))
  ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
test.swift:9:8: note: protocol requires function 'publicMethod' with type '(Float) -> Float'; do you want to add a stub?
  func publicMethod(_ x: Float) -> Float
       ^
```
The last diagnostic is misleading because function `publicMethod` has been declared, but changing/removing it seems non-trivial. I'd like to defer further diagnostic improvements: TF-1154.

rxwei commented:
I was not referring to the attribute-missing diagnostic. I was referring to the differentiation error.
> ```
> test.swift:20:8: error: function is not differentiable
>   func internalMethod(_ x: Float) -> Float {
>        ^~~~~~~~~~~~~~
> test.swift:20:8: note: when differentiating this function definition
>   func internalMethod(_ x: Float) -> Float {
>        ^
> test.swift:22:18: note: cannot differentiate through a non-differentiable result; do you want to use 'withoutDerivative(at:)'?
>     return Float(Int(x))
>                  ^
> ```


This error should show a note that points to the trigger of differentiation, aka. the `@differentiable` attribute on the protocol requirement that got implicitly inherited.

dan-zheng committed:
Use location of inherited `@differentiable` attributes in diagnostics.

Store the location of inherited protocol requirement `@differentiable`
attributes for use in diagnostics.

This improves non-differentiability diagnostics for implicitly inherited
`@differentiable` attributes.

rxwei commented on lib/Sema/TypeCheckProtocol.h:
```diff
@@ -210,9 +210,13 @@ enum class MatchKind : uint8_t {
+  /// The witness is missing a `@differentiable` attribute from the requirement.
+  /// The witness has less-than-public visibility and is declared in a different
+  /// files than the conformance.
+  MissingDifferentiableAttrNonPublicOtherFile,
```
Semantically`MissingDifferentiableAttr` is a superset of `MissingDifferentiableAttrNonPublicOtherFile`. What's the benefit of using two cases compared to using just one?

dan-zheng replied:
Thanks for pointing this out. `MatchKind::MissingDifferentiableAttrNonPublicOtherFile` removed in https://github.com/apple/swift/pull/29771/commits/7d08e9bd8d2a7397b6643f0751530a436328485b.

---

> What's the benefit of using two cases compared to using just one?

In conformance type-checking, requirement-witness mismatches are separately (1) detected (returning a `MatchKind`) and (2) diagnosed (taking a `MatchKind`).

If logic for determining which diagnostic to emit in (1) is sufficiently complicated, it may be more desirable to add a `MatchKind` case rather than duplicating the logic in (2).

In this case, the logic for determining `MatchKind:: MissingDifferentiableAttrNonPublicOtherFile` wasn't complicated (witness is less than public and in separate file from conformance), so duplicating it in (2) makes more sense.

dan-zheng commented:
> This error should show a note that points to the trigger of differentiation, aka. the `@differentiable` attribute on the protocol requirement that got implicitly inherited.

Oops, sorry for the misunderstanding.

I reverted a commit in the wrong direction.
Implicitly inherited `@differentiable` attributes now have a dedicated diagnostic:
```swift
protocol Protocol: Differentiable {
  @differentiable(wrt: (self, x))
  func internalMethod(_ x: Float) -> Float
}

struct ConformingStruct: Protocol {
  func internalMethod(_ x: Float) -> Float {
    return Float(Int(x))
  }
}
```
```console
test.swift:7:8: error: function is not differentiable
  func internalMethod(_ x: Float) -> Float {
       ^~~~~~~~~~~~~~
test.swift:2:4: note: differentiability required by the corresponding protocol requirement here
  @differentiable(wrt: (self, x))
   ^~~~~~~~~~~~~~
test.swift:8:18: note: cannot differentiate through a non-differentiable result; do you want to use 'withoutDerivative(at:)'?
    return Float(Int(x))
                 ^
```

dan-zheng commented on include/swift/AST/Attr.h:
```diff
@@ -1672,6 +1672,13 @@ class DifferentiableAttr final
+  /// This is set during conformance type-checking, only for implicit
+  /// `@differentiable` attributes created for non-public protocol witnesses of
+  /// protocol requirements with `@differentiable` attributes.
+  SourceLoc ImplicitlyInheritedDifferentiableAttrLocation;
```
Note: it's possible to generalize this "diagnostic source location" for other kinds of implicit `@differentiable` attributes too, not just "ones that are implicitly inherited from protocol requirements".

Currently, there are only three kinds of implicit `@differentiable` attributes:
1. This case: `@differentiable` attributes inherited from protocol requirements.
2. `@differentiable` attributes synthesized from others on the same declaration with superset parameter indices.
    - `@differentiable(wrt: (x, y))` -> `@differentiable(wrt: x)`
3. `@differentiable` attributes synthesized on stored properties of structs/classes that derive a conformance to `Differentiable`.
```cpp
enum class ImplicitKind {
  InheritedFromProtocolRequirement, // this case
  SynthesizedSubsetParametersAttribute, // triggered during conformance type-checking
  SynthesizedForStoredProperty, // triggered during `Differentiable` derived conformances
};
```

I think the first case benefits the most from a "secondary diagnostic source location". Generalization could be done later.

dan-zheng committed:
Remove `MatchKind::MissingDifferentiableAttrNonPublicOtherFile`.

rxwei reviewed (approved):
Thanks for addressing all the comments!

dan-zheng commented:
Thanks for the thorough review!
I'll try upstreaming changes to `master` first, then merging the changes into `tensorflow`.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
I accidentally deleted `tensorflow` branch, which closed this PR. That was not intentional, sorry!
It would be nice to protect `tensorflow` branch against deletion to prevent this from happening again.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow
